numba/cuda/tests/cudapy/test_dispatcher.py

```diff
@@ -288,5 +291,42 @@ def add_device(a, b):
         self.assertEqual("Add two integers, device version", add_device.__doc__)
 
 
+@skip_on_cudasim('Dispatcher objects not used in the simulator')
+class TestDispatcherDeprecation(CUDATestCase):
+    def check_warning(self, warnings, expected_str, category):
+        self.assertEqual(len(warnings), 1)
+        for w in warnings:
+            self.assertEqual(w.category, category)
+            self.assertIn(expected_str, str(w.message))
+
+    def test_ptx_deprecation(self):
+        def f(x):
+            x[0] = 42
+
+        with warnings.catch_warnings(record=True) as w:
+            ignore_internal_warnings()
+            warnings.simplefilter("always", category=NumbaDeprecationWarning)
+
+            msg = "Attribute `ptx` is deprecated and will be removed in the "
+            "future. "
+
+            # Kernel code path
+            with self.subTest(name="DispatcherOnCudaKernel"):
```

Comment:
I'm wondering if the warnings list is stateful and as a result these tests could break in strange ways? e.g. should some part of one of these `subTest`s fail such that the `subTest` context is exited early, the warnings list would not be cleared (as `w.clear()` wouldn't be reached) and as a result the next `subTest` would erroneously fail as the warnings list is too long. I think the fix is maybe to copy the statements `with warnings.catch_warnings...` through to the `warnings.simplefilter...` into each subtest, and then to remove the `w.clear()`?

Reply:
Since we will recreate the warning catcher context manager repeatedly, I created a custom ctx manager for re-usability.